pkg/controllers/common/common.go

```diff
@@ -99,3 +139,158 @@ func addConfigMapDigest(cm *corev1.ConfigMap) error {
+		Namespace:   GetNamespace(obj),
+		Annotations: map[string]string{},
+		Labels:      SubResourceLabels(obj),
+	}
+}
+
+func getDiscoverySvcObjMeta(obj client.Object) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:        GetDiscoverySvcName(obj),
+		Namespace:   GetNamespace(obj),
+		Labels:      SubResourceLabels(obj),
+		Annotations: map[string]string{},
+	}
+}
+
+// GetHeadlessService create a headless service
+// https://kubernetes.io/docs/concepts/services-networking/service/#headless-services
+func GetHeadlessService(obj client.Object, ports []corev1.ServicePort) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: getHeadlessSvcObjMeta(obj),
+		Spec: corev1.ServiceSpec{
+			ClusterIP: corev1.ClusterIPNone,
+			Ports:     ports,
+			Selector:  SubResourceLabels(obj),
+		},
+	}
+
+}
+
+// GetDiscoveryService create a service with suffix "-discovery"
+// https://kubernetes.io/docs/concepts/services-networking/service
+func GetDiscoveryService(
+	obj client.Object, ports []corev1.ServicePort, serviceType corev1.ServiceType) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: getDiscoverySvcObjMeta(obj),
+
+		Spec: corev1.ServiceSpec{
+			Type:     serviceType,
+			Ports:    ports,
+			Selector: SubResourceLabels(obj),
+		},
+	}
+}
+
+// GetObjMeta get object metadata
+func GetObjMeta[T client.Object](obj T) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:        GetName(obj),
+		Namespace:   GetNamespace(obj),
+		Annotations: map[string]string{},
+		Labels:      SubResourceLabels(obj),
+	}
+}
+
+func GetConfigMapObjMeta(obj client.Object) metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:        GetConfigName(obj),
+		Namespace:   GetNamespace(obj),
+		Annotations: map[string]string{},
+		Labels:      SubResourceLabels(obj),
+	}
+}
+
+// GetCloneSet get a kruise clone set object
+func GetCloneSet(obj client.Object) *kruise.CloneSet {
+	return &kruise.CloneSet{
+		ObjectMeta: GetObjMeta(obj),
+		Spec: kruise.CloneSetSpec{
+			Selector: &metav1.LabelSelector{
+				MatchLabels: SubResourceLabels(obj),
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: GetObjMeta(obj),
+			},
+		},
+	}
+}
+
+// GetPersistentVolumeClaim return persistent volume claim object
+func GetPersistentVolumeClaim(size resource.Quantity, sc *string) corev1.PersistentVolumeClaim {
+	return corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: DataVolume,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.ResourceRequirements{
+				Requests: map[corev1.ResourceName]resource.Quantity{
+					corev1.ResourceStorage: size,
+				},
+			},
+			StorageClassName: sc,
+		},
+	}
+}
+
+// GetHeadlessSvcName get headless service name
+func GetHeadlessSvcName(obj client.Object) string {
+	return obj.GetName() + hSvcSuffix
+}
+
+// GetDiscoverySvcName get service name
+func GetDiscoverySvcName(obj client.Object) string {
```

Comment:
IMHO this is an implementation detail of logset and we might better not expose it to common